### ISSUE TITLE
Embedding UMAP of functional-region windows (GPN-Star Fig 4) for our gLMs — fold into evals_v2, fwd+RC-averaged center-100 embeddings

### ISSUE THREAD

**Issue by gonzalobenegas** [Open-Athena/marin-dna#246]
🤖 Agent-drafted design/tracking issue. Incorporates decisions made with @gonzalobenegas up front: (1) implement **inside `evals_v2`** (parallel to nuc_dep #237/#238), (2) embed the **last hidden layer only** (match GPN-Star Fig 4; layer kept configurable for a later sweep), and (3) first pass covers **only `exp135-1B-m5.1`** (`mix-v0.9-p1B-i24-exp135-m5.1-step-59158`), config-extensible to more models.

## Goal

Port GPN-Star's **embedding UMAP** interpretation ([`analysis/gpn-star/interpretation`](https://github.com/songlab-cal/gpn/tree/gpn-star-revision/analysis/gpn-star/interpretation), their **Figure 4A/4B**) to our autoregressive marin-dna gLMs: for a balanced set of labeled 100 bp genomic windows (coding, UTRs, lncRNA, promoters, enhancers, background), extract a per-window model embedding, run UMAP, and plot the 2-D projection colored by functional region and by conservation. This is the second interpretation analysis after the nucleotide dependency maps (#237/#238); its dashboard page slots into the **Observatory** Interpretation section (#240), which already reserved a spot for "UMAP / embedding views".

The scientific question (paper §"Visualizing Sequence Embeddings", Fig 4A/4B): **do our models' embeddings segregate functional genomic elements without supervision?** GPN-Star reports that embeddings "largely segregate by genomic region" — CDS and enhancers form distinct clusters, promoter/5′UTR overlap, and **conserved windows cluster more tightly than non-conserved** ones — concluding the model "is aware of core functional elements of the genome." We reproduce the same visualization for our gLMs.

## Background — the embedding UMAP

GPN-Star's recipe (paper Methods + their `interpretation/workflow/scripts/window_embed.py` and `rules/umap.smk`):

1. Take a balanced, repeat-free set of **100 bp** windows (≈ median human coding-exon length), each labeled with a functional region type and a conservation score.
2. **Embed** each window: expand it to the model's context, take per-position embeddings, **average over the center 100 positions**, and **average the forward and reverse-complement** strands → one vector per window.
3. **Standardize** (per-dim z-score) → **UMAP** (defaults, fixed seed) → 2-D.
4. Scatter-plot, colored once by **region label** (Fig 4A) and once by **conservation** (Fig 4B).

The only architecture-specific wrinkle is how step 2 reads embeddings out of the model — GPN-Star is a masked/MSA model; ours are causal `AutoModelForCausalLM` (see "The autoregressive twist" below). Everything else (window set, UMAP, plotting) is reused as-is, so our UMAP is directly comparable to their figure.

## Data — reuse the published region set

We consume GPN-Star's labeled region set **directly** from HuggingFace — [`songlab/gpn-star-umap-regions`](https://huggingface.co/datasets/songlab/gpn-star-umap-regions) — rather than re-deriving it. The dataset *is* their already-sampled, balanced, repeat-excluded window set, so reusing it makes our plot an apples-to-apples reproduction and saves reimplementing the RepeatMasker / Ensembl-113 / ENCODE-SCREEN-v4 / Bioframe sampling pipeline.

<details><summary>Schema — 111,329 windows, GRCh38, single <code>test</code> split (click to expand)</summary>

| Column | Type | Meaning |
|--------|------|---------|
| `chrom` | string | Autosomes `1`–`22`, **no `chr` prefix** (Ensembl-style), GRCh38 |
| `start` | int64 | Window start — **0-based, inclusive** (BED/repo convention) |
| `end` | int64 | Window end — **0-based, exclusive**; `end - start == 100` for every row |
| `cons` | float32 | Conservation = 75th-percentile primate phastCons in the window, ∈ [0, 1] |
| `label` | string | Functional region class (the Fig-4A coloring) |

Label distribution (7 classes, sums to 111,329):

| Label | Count | Display name | Source |
|-------|------:|--------------|--------|
| `CDS` | 20,000 | CDS | Ensembl 113 GFF3 |
| `dELS` | 20,000 | Enhancer | ENCODE SCREEN cCRE v4 |
| `background` | 20,000 | Background | ≥100 bp from any exon/cCRE/repeat |
| `three_prime_UTR` | 18,456 | 3′ UTR | Ensembl 113 GFF3 |
| `lnc_RNA` | 11,991 | lncRNA | Ensembl 113 GFF3 |
| `PLS` | 11,660 | Promoter | ENCODE SCREEN cCRE v4 |
| `five_prime_UTR` | 9,222 | 5′ UTR | Ensembl 113 GFF3 |

The dataset stores **coordinates + label + conservation only — not sequence**; we fetch sequence from our staged GRCh38 reference (`s3://oa-bolinas/data/genomes/homo_sapiens/GRCh38/ensembl-release-115/…dna_sm.primary_assembly.fa.gz`, already used by nuc_dep/compute_scores). GRCh38 ↔ hg38 are assembly-identical and chroms carry no `chr` prefix, matching both the dataset and our `Genome` reader. Coordinates are 0-based half-open — same as the repo convention — so they flow straight into `Genome(chrom, start, end)` with no conversion.

</details>

We use **all 111,329 windows** as-is (the dataset is already the balanced subsample); no further subsampling is needed — embedding extraction is cheap (see Compute) and the plot's point-size/rasterization keeps the SVG small.

### Embedding extraction (the new kernel)

For a window `[start, end)` (exactly 100 bp), center the model's `window_size` context on the window midpoint, run one forward pass with `output_hidden_states=True`, take the **last** layer, mean-pool the center 100 DNA positions, and average forward + reverse-complement:

<details><summary>Indexing detail — center-100 mean-pool with a BOS token (W = 255 for exp135)</summary>

- `center = (start + end) // 2`; `context_start = center - W//2`; fetch `seq = Genome(chrom, context_start, context_start + W, "+")` (length `W`).
- Tokenized as `[BOS, t_0, …, t_{W-1}]` (exp135 has `n_prefix = 1`); DNA position `p` is at token index `n_prefix + p`.
- `hidden_states[-1]` has shape `[1, n_prefix + W, D]`. The **center 100 DNA positions** are `[W//2 - 50, W//2 + 50)`; with `W = 255` that is `[77, 177)` — token indices `[78, 178)` — which is **exactly the dataset's `[start, end)` window** (we centered the context on it). Mean-pool those 100 → `[D]`.
- Reverse-complement pass: `reverse_complement(seq)`, same extraction, mean-pool the (symmetric) center 100 → `[D]`. No coordinate flip is needed (unlike nuc_dep): we mean-pool a symmetric central block, so position order within the pool is irrelevant.
- Per-window embedding = `(fwd + rc) / 2`, shape `[D]`.

For throughput the kernel batches many windows per forward pass (2 rows per window for fwd+RC). Windows whose expanded context runs off a chromosome end or hits an assembly gap come back with `N`; we **drop** those windows and assert the dropped fraction is tiny (these are bulk data, not the curated handful nuc_dep hard-asserts on).

</details>

### The autoregressive twist (lighter than nuc_dep)

GPN-Star is bidirectional, so its center-100 embedding sees both flanks. Our models are **causal**: the hidden state at position `p` encodes only `x_{≤p}` (left context), so a single-strand center-100 mean-pool is left-context-biased. Averaging the **forward and reverse-complement** embeddings symmetrizes this — each strand contributes the opposite-direction context — analogous to (but much milder than) the FWD/RC *triangle stitch* in nuc_dep. Here it is just a mean of two `[D]` vectors. The paper averages fwd+RC too (for them it is strand-invariance / variance reduction; for us it additionally corrects the causal left-bias), so we stay aligned with the reference while doing the right thing for a causal LM.

### UMAP + plotting (reuse the reference recipe verbatim)

<details><summary>UMAP and plot parameters (matching GPN-Star)</summary>

- **UMAP**: `sklearn` `StandardScaler` → `umap.UMAP(random_state=42)` (everything else default: `n_neighbors=15`, `min_dist=0.1`, `metric="euclidean"`, `n_components=2`). No PCA pre-reduction. (`random_state` forces single-threaded UMAP — fine at this scale.)
- **Plot** (seaborn scatter, two SVGs per model — one `hue=label`, one `hue=cons`): `figsize=(3, 3)`, point size `s = 100/√N`, `alpha=0.5`, `linewidth=0`, `rasterized=True`, save at `dpi=200` with `bbox_inches="tight"`. Region palette uses the 7 display names above with a fixed color per class; conservation uses a continuous map. Legend moved outside, `despine()`, no ticks. Emit `.svg` (dashboard artifact) **and** a `.png` sibling (repo convention — for `Read`-back sanity checks). The `s = 100/√N` + rasterized-points trick is exactly how GPN-Star keeps a ~100 k-point SVG small.

</details>

## Where it lives — fold into `evals_v2` (parallel to nuc_dep)

Mirror the nuc_dep #237/#238 structure: thin Snakemake rules over library functions, kept **off the default `rule all`** so they never perturb metric reruns. Build explicitly (`snakemake umap`).

- **Library kernel** — `src/marin_dna/model/embeddings.py`: `window_embeddings(model, tokenizer, seqs, *, layer_index=-1, n_center_bp=100, rc=True, batch_size=…) -> ndarray[N, D]`. Reuses `_get_special_token_counts` (BOS count) and `reverse_complement`; duck-typed model interface (`model(input_ids, output_hidden_states=True).hidden_states[layer_index]`), same as `scoring.py`/`interpretation.py`.
- **Orchestration** — `src/marin_dna/pipelines/evals/embedding_umap.py`: `compute_region_embeddings(checkpoint, genome, regions_df, window_size, …)` (load model, fetch sequence per window, batch, drop-N, return embeddings + carried `chrom/start/end/label/cons`), `fit_umap(emb_df, …)`, and `plot_umap(umap_df, out, color_by, …)`.
- **Dashboard catalog** — extend `src/marin_dna/pipelines/evals/interpretation_catalog.py` (reuse `S3_BUCKET`, model-display map) with UMAP candidates + an S3 plot prefix.
- **Snakemake** — new `snakemake/analysis/evals_v2/workflow/rules/embedding_umap.smk` (included after `interpretation.smk`), with a new `umap_embeddings:` config block validated in `common.smk` (model exists; `window_size ≥ 100`). Rules:
  - `compute_region_embeddings` (GPU) → `results/interpretation/umap/{model}/embeddings.parquet`
  - `fit_umap` (CPU) → `results/interpretation/umap/{model}/umap_coords.parquet`
  - `plot_umap` (CPU) → `results/plots/umap/{model}/region.svg` + `…/conservation.svg`
  - `umap` aggregator (not in `rule all`)
- **Dashboard** — `dashboard/src/data/umap.zip.py` loader (bundle the SVGs + `manifest.json`, mirroring `nuc_dep.zip.py`) and a `dashboard/src/interpretation/embedding-umap.md` page in the Observatory nav (#240).
- **New dependency — optional group, not core.** `umap-learn` pulls in `numba` + `llvmlite` (a **~56 MB LLVM wheel**, ~60 MB total) for this single analysis, so it goes in a new **`[dependency-groups]` `umap`** group (alongside `genome-s3` / `alphagenome-eval` / `enhancer-classification`), *not* core deps — every training/eval/CI env would otherwise carry the LLVM wheel for nothing. `fit_umap` **lazy-imports** `umap` inside the function (mirroring how `plot_dependency_map` imports matplotlib/seaborn), so the rest of the library and pytest *collection* never touch it; the one umap test is guarded with `pytest.importorskip("umap")` (CI's `dev` group stays LLVM-free). The pipeline env opts in via `uv sync --frozen --group genome-s3 --group umap` (the compute rule already needs `genome-s3` for S3 sequence reads). `datasets` / `polars` / `scikit-learn` / `seaborn` / `matplotlib` are already core.

<details><summary>Proposed <code>umap_embeddings:</code> config block</summary>

```yaml
umap_embeddings:
  dataset: songlab/gpn-star-umap-regions   # HF dataset id (test split)
  layer_index: -1                          # last hidden layer (match GPN-Star); configurable for a sweep
  n_center_bp: 100                         # center positions to mean-pool
  batch_size: 64                           # windows per forward pass (execution-only; tune for VRAM)
  dpi: 200                                 # scatter raster dpi
  models:
    - mix-v0.9-p1B-i24-exp135-m5.1-step-59158   # exp135-1B-m5.1 (255+BOS) — first pass, single model
```

</details>

## Compute plan

Small GPU, like nuc_dep. exp135-1B over 111,329 windows × 2 strands = **222,658** forward passes of ~255 tokens — batched on one A10G / `g5.xlarge` this is on the order of **15–40 min**; UMAP on ~111 k × D points (single-threaded) adds a few minutes. Reuse the existing `sky/run.yaml` + `$SNAKEMAKE_ARGS` pattern (its setup adds `--group umap` to the existing `--group genome-s3`), target = the two plot SVGs. **Will confirm before any SkyPilot launch** (per repo policy); if a GPU is already up locally we can validate there first.

## Validation & sanity checks

- Shape/finiteness: embedding matrix `[N, D]`, no NaNs; assert center-100 slice indices in-bounds and `n_prefix ≥ 1` honored.
- N-drop accounting: log and assert the dropped (off-chromosome / assembly-gap) fraction is small (< ~0.5%).
- Qualitative reproduction of Fig 4: CDS and enhancers (`dELS`) form distinct clusters; promoter (`PLS`)/5′UTR overlap; conserved windows cluster more tightly (color-by-`cons` panel). Eyeball the PNG.
- (Optional quantitative) a region-separation score on the embeddings — e.g. LOO 1-NN balanced accuracy or silhouette by `label` — as a single number complementing the eyeball. This dovetails with the deferred Consens region-separation eval interest (#234) and would let us compare layers/models objectively.

## Decisions recorded (from this brainstorm)

- ✅ **Last hidden layer only** (match GPN-Star Fig 4); `layer_index` configurable for a later sweep.
- ✅ **First pass = `exp135-1B-m5.1` only** (`mix-v0.9-p1B-i24-exp135-m5.1-step-59158`, 255+BOS); config-extensible.
- ✅ **Reuse the HF dataset** `songlab/gpn-star-umap-regions` directly (no re-derivation of the region set); all 111,329 windows, no further subsampling.
- ✅ **Fold into `evals_v2`**, off `rule all`; reuse `download_model` / `genome_path` / S3 storage / `sky/run.yaml`.
- ✅ **Two plots per model**: by region label (Fig 4A) and by conservation (Fig 4B).
- ✅ **fwd+RC averaged, standardized, UMAP defaults + `random_state=42`**.
- ✅ **`umap-learn` is an optional `uv` dependency-group** (`umap`), lazy-imported in `fit_umap` — *not* a core dep (it pulls a ~56 MB LLVM wheel via `numba`/`llvmlite`).

## Deferred / extensions

- **Layer sweep** — `output_hidden_states=True` yields every layer from the *same* forward pass; emit a few (final / ~¾ / ~½ depth), UMAP+plot each, feature the best-segregating. Near-free GPU-wise.
- **More models** — exp166-v0.1-p1B and the no-BOS arms (exp55/58/59; embeddings don't require BOS), other sizes/families, for a matched dashboard comparison.
- **Quantitative region-separation metric** (1-NN balanced accuracy / silhouette / linear probe) — the dataset is literally pitched as a region-classification benchmark; ties to #234.
- **Published-model baseline** — a single-sequence reference (GPN / Nucleotide Transformer) side-by-side; a true GPN-Star reproduction needs their multiz/cactus MSAs, which is heavier infra and out of scope here.

## References

- **Paper**: Ye et al., "GPN-Star", bioRxiv [2025.09.21.677619](https://doi.org/10.1101/2025.09.21.677619) — Fig 4A/4B; Methods §"Visualizing Sequence Embeddings".
- **Reference code**: [`songlab-cal/gpn @ gpn-star-revision` `analysis/gpn-star/interpretation`](https://github.com/songlab-cal/gpn/tree/gpn-star-revision/analysis/gpn-star/interpretation) (`rules/umap.smk`, `scripts/window_embed.py`).
- **Dataset**: [`songlab/gpn-star-umap-regions`](https://huggingface.co/datasets/songlab/gpn-star-umap-regions).
- **Our precedent**: nucleotide dependency maps #237 / #238; Observatory dashboard #240.

**Comment by gonzalobenegas** [comment]
🤖 Implemented in #248 (branch `claude/issue-246-embedding-umap`). The exp135 run reproduces GPN-Star Fig 4A/4B — CDS and enhancers form distinct clusters, and those are the high-conservation ones (figures in the PR).

Two deliberate departures from the design drafted above, decided during implementation (see the PR commits):
- **Embedding runs on the shared HF `Trainer` harness** (bf16 / `torch.compile` / dataloader workers, base `AutoModel`) rather than a hand-rolled loop — ~3.2× less VRAM.
- **Allow-N instead of drop-N**: every one of the 111,329 windows is embedded (`N` in the flanks → `[UNK]`; the labeled center is always real), so the point set is identical across models of different context sizes and the UMAPs stay point-for-point comparable. The design notes above (drop-N, custom loop) are superseded by the PR.
